File: ports/go/src/core/kernel/bcsobject.go
package kernel

import (
	"sync"
)

// BcsObject forms the base for all framework components in the Go port, mapping to C++ QObject
type BcsObject struct {
	mu           sync.Mutex
	onDestroy    func()
	parent       *BcsObject
	children     []*BcsObject
	objectName   string
	dispatcher   *BcsEventDispatcher
	isDestroyed  bool
}

func NewBcsObject(parent *BcsObject) *BcsObject {
	obj := &BcsObject{
		parent:     nil, // We set this via SetParent to ensure correct initial linking
		children:   make([]*BcsObject, 0),
		dispatcher: nil,
	}

	if parent != nil {
		obj.SetParent(parent)
	}
	return obj
}

func (o *BcsObject) SetObjectName(name string) {
	o.mu.Lock()
	defer o.mu.Unlock()
	o.objectName = name
}

func (o *BcsObject) ObjectName() string {
	o.mu.Lock()
	defer o.mu.Unlock()
	return o.objectName
}

func (o *BcsObject) Parent() *BcsObject {
	o.mu.Lock()
	defer o.mu.Unlock()
	return o.parent
}

// addChild is an internal helper that must be called with the parent's lock already held.
func (o *BcsObject) addChild(child *BcsObject) {
	for _, c := range o.children {
		if c == child {
			return
		}
	}
	o.children = append(o.children, child)
}

// removeChild is an internal helper that must be called with the parent's lock already held.
func (o *BcsObject) removeChild(child *BcsObject) {
	for i, c := range o.children {
		if c == child {
			o.children[i] = o.children[len(o.children)-1]
			o.children = o.children[:len(o.children)-1]
			return
		}
	}
}

func (o *BcsObject) SetOnDestroy(f func()) {
	o.mu.Lock()
	defer o.mu.Unlock()
	o.onDestroy = f
}

func (o *BcsObject) SetParent(newParent *BcsObject) {
	// First, acquire our own lock to detach from old parent
	o.mu.Lock()
	oldParent := o.parent
	o.parent = newParent
	o.mu.Unlock()

	// Carefully detach from old parent using its internal method
	if oldParent != nil && oldParent != newParent {
		oldParent.mu.Lock()
		oldParent.removeChild(o)
		oldParent.mu.Unlock()
	}

	// Carefully attach to new parent
	if newParent != nil {
		newParent.mu.Lock()
		newParent.addChild(o)
		newParent.mu.Unlock()
	}
}

// Destroy cleans up the object tree recursively (simulating C++ destructor)
func (o *BcsObject) Destroy() {
	o.mu.Lock()
	if o.isDestroyed {
		o.mu.Unlock()
		return
	}
	o.isDestroyed = true

	// Call the subclass destructor hook if defined
	if o.onDestroy != nil {
		o.onDestroy()
	}

	// Copy children to avoid deadlock during destruction
	childrenCopy := make([]*BcsObject, len(o.children))
	copy(childrenCopy, o.children)
	o.mu.Unlock()

	for _, child := range childrenCopy {
		child.Destroy()
	}

	o.SetParent(nil)
	o.children = nil
}

// Event receives events from the dispatcher
func (o *BcsObject) Event(e BcsEvent) bool {
	// Virtual function mapping. Subclasses handle specific events.
	if e.Type() == Timer {
		o.TimerEvent(e.(*BcsTimerEvent))
		return true
	} else if e.Type() == ChildAdded || e.Type() == ChildRemoved {
		o.ChildEvent(e.(*BcsChildEvent))
		return true
	}
	return false
}

// Event filters
func (o *BcsObject) TimerEvent(e *BcsTimerEvent) {}
func (o *BcsObject) ChildEvent(e *BcsChildEvent) {}

File: ports/rust/src/core/kernel/bcs_object.rs
use std::sync::{Arc, Mutex, Weak};
use super::bcs_event::{BcsEventT, EventType};

/// BcsObject provides the object tree hierarchy mapping to QObject
pub struct BcsObject {
    pub object_name: Mutex<String>,
    pub parent: Mutex<Option<Weak<BcsObject>>>,
    pub children: Mutex<Vec<Arc<BcsObject>>>,
    pub is_destroyed: Mutex<bool>,
    // Store a self weak reference to allow detached parent mutation
    pub _self_weak: Mutex<Option<Weak<BcsObject>>>,
}

impl BcsObject {
    pub fn new() -> Arc<Self> {
        let obj = Arc::new(Self {
            object_name: Mutex::new(String::new()),
            parent: Mutex::new(None),
            children: Mutex::new(Vec::new()),
            is_destroyed: Mutex::new(false),
            _self_weak: Mutex::new(None),
        });
        *obj._self_weak.lock().unwrap() = Some(Arc::downgrade(&obj));
        obj
    }

    pub fn new_with_parent(parent: Arc<BcsObject>) -> Arc<Self> {
        let obj = Arc::new(Self {
            object_name: Mutex::new(String::new()),
            parent: Mutex::new(Some(Arc::downgrade(&parent))),
            children: Mutex::new(Vec::new()),
            is_destroyed: Mutex::new(false),
            _self_weak: Mutex::new(None),
        });
        *obj._self_weak.lock().unwrap() = Some(Arc::downgrade(&obj));

        parent.add_child(obj.clone());
        obj
    }

    pub fn set_object_name(&self, name: &str) {
        *self.object_name.lock().unwrap() = name.to_string();
    }

    pub fn object_name(&self) -> String {
        self.object_name.lock().unwrap().clone()
    }

    pub fn parent(&self) -> Option<Arc<BcsObject>> {
        if let Some(weak_p) = &*self.parent.lock().unwrap() {
            return weak_p.upgrade();
        }
        None
    }

    pub fn set_parent(self: &Arc<Self>, parent: Option<Arc<BcsObject>>) {
        let old_parent = self.parent();

        *self.parent.lock().unwrap() = parent.as_ref().map(|p| Arc::downgrade(p));

        if let Some(old) = old_parent {
            if let Some(new_p) = &parent {
                if !Arc::ptr_eq(&old, new_p) {
                    old.remove_child(self);
                }
            } else {
                old.remove_child(self);
            }
        }

        if let Some(new_p) = parent {
            new_p.add_child(self.clone());
        }
    }

    pub fn add_child(&self, child: Arc<BcsObject>) {
        let mut children = self.children.lock().unwrap();
        // Prevent duplicates
        if !children.iter().any(|c| Arc::ptr_eq(c, &child)) {
            children.push(child);
        }
    }

    pub fn remove_child(&self, child: &Arc<BcsObject>) {
        let mut children = self.children.lock().unwrap();
        if let Some(pos) = children.iter().position(|c| Arc::ptr_eq(c, child)) {
            children.remove(pos);
        }
    }

    pub fn destroy(&self) {
        let mut destroyed = self.is_destroyed.lock().unwrap();
        if *destroyed {
            return;
        }
        *destroyed = true;

        let children_copy = {
            let mut children = self.children.lock().unwrap();
            let copy = children.clone();
            children.clear();
            copy
        };

        for child in children_copy {
            child.destroy();
        }

        // Fix: Properly remove self from parent's children vector using the _self_weak ref
        if let Some(p) = self.parent() {
            if let Some(self_arc) = self._self_weak.lock().unwrap().as_ref().and_then(|w| w.upgrade()) {
                p.remove_child(&self_arc);
            }
        }
        *self.parent.lock().unwrap() = None;
    }

    pub fn event(&self, _event: &dyn BcsEventT) -> bool {
        // Base implementation does nothing
        false
    }
}
